### PR TITLE
Refactor middle

### PR DESCRIPTION
**Pull request by joto**
Various refactorings of the middle, removing the slim_middle_t class etc.

### CODE REVIEW

File: src/osmdata.cpp [hunk 1]
```diff
@@ -34,21 +34,12 @@ osmdata_t::osmdata_t(std::unique_ptr<dependency_manager_t> dependency_manager,
...
-slim_middle_t &osmdata_t::slim_middle() const noexcept
-{
-    auto *slim = dynamic_cast<slim_middle_t *>(m_mid.get());
-    assert(slim);
-    return *slim;
-}
-
 void osmdata_t::node(osmium::Node const &node)
 {
+    if (!m_bbox.valid() || node.deleted() || m_bbox.contains(node.location())) {
```
Comment (lonvia) [inline src/osmdata.cpp:39]: contains() is now called twice. Not the cheapest of functions.

More importantly: you either need to move the node.location().valid() check to the front of the function or make sure that mid can handle invalid locations. I'd be in favour of the first solution. (Bonus points for adding a test for input with invalid locations.)

File: src/middle-pgsql.cpp [hunk 2]
```diff
@@ -314,6 +314,46 @@ std::size_t middle_query_pgsql_t::get_way_node_locations_db(
...
+void middle_pgsql_t::node(osmium::Node const &node)
+{
+    if (node.deleted()) {
+        node_delete(node.id());
+    } else {
+        if (m_options->append) {
+            node_delete(node.id());
+            node_set(node);
+        } else {
+            node_set(node);
```
Comment (lonvia) [inline src/middle-pgsql.cpp:327]: Any particular reason why you don't move the node_set() outside the if? Dto for the other two functions.

Comment (joto) [reply src/middle-pgsql.cpp]: Just not gotten around to that. I have added a commit now that does that.

File: src/osmdata.cpp (hunk 1 above)
Comment (joto) [reply src/osmdata.cpp]: In an additional commit I have refactored this and added a test.

**Pull request merged**